### PR TITLE
Remove usage of `purge_nonempty_nulls` in set operations

### PR DESCRIPTION
**Pull request by ttnghia**
As `purge_nonempty_nulls` is implicitly called in `cudf::make_lists_column` (https://github.com/rapidsai/cudf/pull/12370) and it is only called if absolutely needed (https://github.com/rapidsai/cudf/pull/12547), now we can remove usage of that API altogether in set operations. This may also potential speed up set operation APIs to some extent.

### CODE REVIEW

**Comment by ttnghia** [comment]
I also added a benchmark. The numbers show (significant) improvement in set ops performance if having nulls:
```
# intersect_distinct

## [0] Quadro RTX 6000

|  num_rows  |  depth  |  null_frequency  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |  Status  |
|------------|---------|------------------|------------|-------------|------------|-------------|---------------|---------|----------|
|    2^10    |    1    |        0         | 259.136 us |      31.50% | 268.791 us |      36.01% |      9.655 us |   3.73% |   PASS   |
|    2^13    |    1    |        0         | 291.603 us |      16.24% | 288.337 us |      14.62% |     -3.266 us |  -1.12% |   PASS   |
|    2^16    |    1    |        0         | 921.890 us |       6.64% | 918.140 us |       5.94% |     -3.750 us |  -0.41% |   PASS   |
|    2^10    |    4    |        0         |  25.387 ms |       4.92% |  25.312 ms |       3.78% |    -74.761 us |  -0.29% |   PASS   |
|    2^13    |    4    |        0         | 134.174 ms |       1.72% | 134.329 ms |       0.49% |    154.267 us |   0.11% |   PASS   |
|    2^16    |    4    |        0         | 416.889 ms |       1.35% | 417.986 ms |       0.49% |      1.097 ms |   0.26% |   PASS   |
|    2^10    |    1    |       0.2        | 381.686 us |      25.54% | 287.477 us |       7.96% |    -94.208 us | -24.68% |   FAIL   |
|    2^13    |    1    |       0.2        | 401.864 us |      13.03% | 323.630 us |       8.42% |    -78.234 us | -19.47% |   FAIL   |
|    2^16    |    1    |       0.2        | 839.671 us |      11.39% | 747.388 us |       4.95% |    -92.283 us | -10.99% |   FAIL   |
|    2^10    |    4    |       0.2        |  27.860 ms |       1.70% |  27.683 ms |       5.11% |   -176.944 us |  -0.64% |   PASS   |
|    2^13    |    4    |       0.2        |  30.731 ms |       2.47% |  27.856 ms |       3.08% |  -2874.663 us |  -9.35% |   FAIL   |
|    2^16    |    4    |       0.2        | 244.873 ms |       1.42% | 220.665 ms |       1.40% | -24207.715 us |  -9.89% |   FAIL   |
|    2^10    |    1    |       0.8        | 383.869 us |      28.14% | 296.832 us |      28.36% |    -87.036 us | -22.67% |   PASS   |
|    2^13    |    1    |       0.8        | 374.298 us |      11.99% | 294.892 us |      15.78% |    -79.406 us | -21.21% |   FAIL   |
|    2^16    |    1    |       0.8        | 387.524 us |       7.28% | 313.615 us |      17.14% |    -73.908 us | -19.07% |   FAIL   |
|    2^10    |    4    |       0.8        |   3.120 ms |       7.52% |   2.679 ms |      13.28% |   -440.774 us | -14.13% |   FAIL   |
|    2^13    |    4    |       0.8        |   3.495 ms |       6.76% |   2.899 ms |       6.64% |   -596.671 us | -17.07% |   FAIL   |
|    2^16    |    4    |       0.8        |   3.892 ms |       7.75% |   3.253 ms |       7.95% |   -639.258 us | -16.42% |   FAIL   |

```

**Review by divyegala — APPROVED** [review]

File: cpp/tests/lists/set_operations/intersect_distinct_tests.cpp [hunk 1]
```diff
@@ -280,11 +281,18 @@ TEST_F(SetIntersectTest, StringTestsWithNullsUnequal)
...
       null_at(1)};
-    auto const expected =
-      strings_lists{{strings_lists{}, strings_lists{} /*NULL*/, strings_lists{"this"}}, null_at(1)};
+    auto const expected = [] {
+      auto str_lists =
+        strings_lists{{strings_lists{}, strings_lists{} /*NULL*/, strings_lists{"this"}},
+                      null_at(1)}
+          .release();
+      auto& child = str_lists->child(cudf::lists_column_view::child_column_index);
+      child.set_null_mask(cudf::create_null_mask(child.size(), cudf::mask_state::ALL_VALID), 0);
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/tests/lists/set_operations/intersect_distinct_tests.cpp:290]: This is due to `cudf::copy_if` used internally that creates all valid null mask, so we have to adapt the unit tests accordingly.

[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/tests/lists/set_operations/intersect_distinct_tests.cpp:290]: Previously we called purge nonempty nulls which removed such all valid null mask.

**Comment by ttnghia** [comment]
Taking this into draft as this may not be needed anymore, since we may change our strategy to deal with list sanitization:
 * If libcudf algorithms are generating unsanitized columns, fix the algorithms.

**Comment by ttnghia** [comment]
Okay, this is ready for review again: it is still relevant.

**Pull request closed without merging**